server/repository/src/db_diesel/changelog/changelog.rs

```diff
@@ -464,6 +471,52 @@ fn create_filtered_outgoing_sync_query(
     query
 }
 
+/// Runs some DB operation with a fully locked `changelog` table.
+/// This only applies for for Postgres and does nothing for Sqlite.
+///
+/// Motivation:
+/// When querying changelog entries, ongoing transactions might continue adding changelog entries
+/// to the queried range of changelogs.
+/// This is because Postgres has Read Committed isolation level (instead of Serialized in Sqlite).
+/// However, we assume that there will be no new changelog entries in the queried range in the
+/// future, e.g. when updating the cursor position.
+///
+/// For example, a changelog may contain [1, 3, 4, 5] while another (slow) tx is about to commit a
+/// changelog row with cursor = 2.
+/// We need to wait for this changelog 2 to be added before doing the changelogs() query, otherwise
+/// we might update the latest changelog cursor to 5 and the changelog with cursor = 2 will be left
+/// unhandled when continuing from the latest cursor position.
+///
+/// Locking the changelog table will wait for ongoing writers and will prevent new writers while
```

Comment:
While there is a lock on a table, what happens to existing transactions that try to write to the table (what does `will prevent` mean) ?

Reply:
I think the database will wait to try to acquire the lock (until a wait timeout). It's not only writing to the table, any reads will be blocked too.

Reply:
As I understood other transactions will wait till the lock is released.
And the read transaction has to wait for other transactions before it can acquire the lock.

Btw. I run the transfer tests 50 times in a row without error. (However might not mean much if there is an even subtler bug, just an indication that I reasonably tested it...)

Reply:
Existing transactions can keep writing their data.

Reply:
@jmbrunskill 

> Existing transactions can keep writing their data.

What do you mean ?

Reply:
I just tested it and it appears to work like this...

Let's imagine we have a transaction which need to write `invoice` and `invoice_line` to `changelog` in progress.
The transaction writes it's `invoice` changelog record, then a change log read request asks for exclusive mode (Transaction 2)

Transaction 2 will wait until transaction 1 is finished. It is able to insert it's remaining changelog records as it already holds a write lock on the table, and transaction 2 can't claim it's lock until all the other locks are cleared.

Transaction 3 which wants to start writing to change log, wouldn't be able to start until transaction 2 is finished.